services/src/main/java/org/keycloak/protocol/ciba/decoupledauthn/DecoupledAuthenticationProviderBase.java

```diff
@@ -0,0 +1,230 @@
+        // here Client Model of CD(Consumption Device) needs to be used to bind its Client Session with User Session.
+        AuthenticationSessionModel authSession = rootAuthSession.createAuthenticationSession(client);
+
+        authSession.setProtocol(OIDCLoginProtocol.LOGIN_PROTOCOL);
+        authSession.setAction(AuthenticatedClientSessionModel.Action.AUTHENTICATE.name());
+        authSession.setClientNote(OIDCLoginProtocol.ISSUER, Urls.realmIssuer(session.getContext().getUri().getBaseUri(), realm.getName()));
+        authSession.setClientNote(OIDCLoginProtocol.SCOPE_PARAM, getScope());
+
+        // authentication
+        AuthenticationFlowModel flow = AuthenticationFlowResolver.resolveCIBAFlow(authSession);
+        String flowId = flow.getId();
+        AuthenticationProcessor processor = new AuthenticationProcessor();
+        processor.setAuthenticationSession(authSession)
+                .setFlowId(flowId)
+                .setConnection(clientConnection)
+                .setEventBuilder(event)
+                .setRealm(realm)
+                .setSession(session)
+                .setUriInfo(session.getContext().getUri())
+                .setRequest(httpRequest);
+
+        processor.authenticateOnly();
+        processor.evaluateRequiredActionTriggers();
+        UserModel user = authSession.getAuthenticatedUser();
+        if (user.getRequiredActionsStream().count() > 0) {
+            event.error(Errors.RESOLVE_REQUIRED_ACTIONS);
+            throw new CorsErrorResponseException(cors, OAuthErrorException.INVALID_GRANT, "Account is not fully set up", Response.Status.BAD_REQUEST);
+        }
+
+        AuthenticationManager.setClientScopesInSession(authSession);
+
+        processor.attachSession();
+        UserSessionModel userSession = processor.getUserSession();
+        if (userSession == null) {
+            event.error(Errors.USER_SESSION_NOT_FOUND);
+            throw new CorsErrorResponseException(cors, OAuthErrorException.INVALID_GRANT, "User session is not found", Response.Status.BAD_REQUEST);
+        }
+        userSession.getNotes().putAll(clientAuthAttributes);
+        logger.tracef("CIBA Grant :: specified scopes in backchannel authentication endpoint = %s, Created User Session's id = %s, Submitted in advance User Session ID Will Be Created = %s, username = %s", getScope(), userSession.getId(), getUserSessionIdWillBeCreated(), userSession.getUser().getUsername());
+
+        // authorization (consent)
+        UserConsentModel grantedConsent = session.users().getConsentByClient(realm, user.getId(), client.getId());
+        if (grantedConsent == null) {
+            grantedConsent = new UserConsentModel(client);
+            session.users().addConsent(realm, user.getId(), grantedConsent);
+            grantedConsent.getGrantedClientScopes().forEach(i->logger.tracef("CIBA Grant :: Consent granted. %s", i.getName()));
```

Comment:
I would add an if statement prior to running this line to check if trace level is enabled.

Reply:
@pedroigor Yes, it is rational. I will add trace log level checks before logging.